File: core/triagem_email.py
"""Triagem do e-mail na página inicial: soltar o pedido e cair na tela certa.

A pessoa solta o e-mail (ou o processo do eProtocolo em PDF) na faixa da
Central de módulos. `ler` lê a mensagem (`core.leitura.mensagem`), guarda o
original na pasta temporária sob o módulo "triagem" e responde com os
módulos candidatos (`core.preencher_por_email.candidatos_de_triagem`, que
soma os sinais do texto e a memória dos pedidos anteriores). Com um
candidato claro, o navegador vai direto para a tela nova daquele módulo com
`?email_origem=<token>`; a tela lê o e-mail sozinha ao abrir
(`origem_da_tela` + `data-pe-auto`) e oferece os outros módulos. Sem
candidato claro, a faixa mostra as opções para a pessoa escolher.

`encaminhar` é o clique numa opção: confere que o token é desta sessão e
manda para a tela nova do módulo escolhido.

Nada é gravado: o e-mail só vira registro quando o formulário é salvo, e o
que se aprende dele (`core.aprendizado`) é registrado nesse momento.
"""

from __future__ import annotations

import logging

from django.contrib.auth.decorators import login_required
from django.http import HttpResponseBadRequest, JsonResponse
from django.shortcuts import redirect
from django.utils import timezone
from django.views.decorators.http import require_POST

from core import preencher_por_email as pe
from core.leitura.mensagem import Mensagem

logger = logging.getLogger(__name__)

#: Confiança a partir da qual a página inicial abre a tela sem perguntar.
CONFIANCA_PARA_DECIDIR = 0.55


def _resumo(mensagem: Mensagem) -> dict:
    """O que a faixa mostra sobre o e-mail lido, antes de abrir a tela."""
    enviado = mensagem.enviado_em
    if enviado is not None and timezone.is_aware(enviado):
        enviado = timezone.localtime(enviado)
    quando, avisos = pe.quando_do_pedido(mensagem)
    pessoa = pe.quem_pede(mensagem)
    return {
        "assunto": mensagem.assunto_limpo[:200],
        "remetente": (mensagem.remetente or "")[:150],
        "enviado_em": f"{enviado:%d/%m/%Y %H:%M}" if enviado else "",
        "quando": f"{quando.inicio:%d/%m/%Y}" if quando else "",
        "quem": (pessoa.nome if pessoa else "")[:150],
        "avisos": [a for a in list(mensagem.avisos) + avisos if a][:4],
        "origem": mensagem.origem,
    }


@login_required
@require_POST
def ler(request):
    """Lê o e-mail solto na página inicial e diz para qual módulo ele parece ir."""
    try:
        mensagem, nome, dados = pe.ler_do_pedido(request)
    except pe.EmailRecusado as erro:
        return JsonResponse({"erro": str(erro)}, status=400)
    token = pe._guardar(request, pe.MODULO_TRIAGEM, nome, dados, mensagem)
    # Os candidatos com sinal e, depois, os outros módulos da pessoa (sem sinal, ela escolhe).
    opcoes = pe.opcoes_de_triagem(mensagem, request.user, token)
    sugerido = opcoes[0] if opcoes and opcoes[0]["confianca"] else None
    decidir = bool(sugerido) and sugerido["confianca"] >= CONFIANCA_PARA_DECIDIR
    logger.info(
        "Triagem de e-mail: origem=%s, %d opção(ões), sugerido=%s, decidir=%s.",
        mensagem.origem, len(opcoes), sugerido["modulo"] if sugerido else "-", decidir,
    )
    return JsonResponse({
        "token": token,
        "arquivo": nome,
        "resumo": _resumo(mensagem),
        "candidatos": opcoes,
        "sugerido": sugerido["modulo"] if sugerido else "",
        "decidir": decidir,
    })


@login_required
def encaminhar(request):
    """Abre a tela nova do módulo escolhido com o e-mail lido na página inicial."""
    token = (request.GET.get("token") or "").strip().lower()
    modulo = (request.GET.get("modulo") or "").strip()
    if modulo not in pe.MODULOS_TRIAGEM or modulo not in pe.modulos_de_triagem(request.user):
        return HttpResponseBadRequest("Módulo inválido.")
    if not pe.reencaminhar(request, token, modulo):
        return redirect("core:home")
    return redirect(f"{pe.url_da_tela_nova(modulo)}?email_origem={token}")

File: static/js/triagem-email.js
/**
 * Triagem do e-mail na página inicial (components/v32/triagem_email.html).
 *
 * A pessoa solta o e-mail (ou o processo do eProtocolo em PDF) ou cola o
 * texto. O endpoint `core:triagem_ler` (POST) responde
 *   {token, resumo: {assunto, remetente, enviado_em, quando, quem, avisos},
 *    candidatos: [{modulo, rotulo, confianca, sinais, url}], sugerido, decidir}
 * e aqui:
 *
 * - com `decidir`, a página vai direto para a tela nova do módulo sugerido
 *   (`url`, com ?email_origem=<token>): lá a faixa "Preencher com um e-mail"
 *   lê o e-mail sozinha e oferece os outros módulos;
 * - sem, mostra o resumo do e-mail e um botão por módulo candidato (o
 *   sugerido em destaque) para a pessoa escolher.
 *
 * O conteúdo do e-mail nunca entra como HTML: tudo vai por textContent.
 */
(function () {
  "use strict";

  var EXTENSOES = [".eml", ".msg", ".pdf", ".txt"];

  // O celular entrega o arquivo como uma referência que às vezes expira antes
  // do envio (Drive, Gmail, "Recentes") — o fetch morre com "Failed to fetch".
  // Ler o conteúdo já, na hora da escolha, e mandar a cópia evita isso; se
  // nem a leitura der, a mensagem explica o que fazer.
  var MSG_LEITURA = "Não consegui abrir o arquivo no aparelho. Salve-o no telefone (pasta Downloads) e escolha de novo pelo app Arquivos — ou use o computador.";
  var MSG_REDE = "A conexão caiu no envio do arquivo. Confira a internet e tente de novo; se continuar, salve o arquivo no aparelho e escolha de novo.";
  function copiaDe(arquivo) {
    if (!arquivo.arrayBuffer) return Promise.resolve(arquivo);
    return arquivo.arrayBuffer().then(function (conteudo) {
      return new File([conteudo], arquivo.name || "arquivo", { type: arquivo.type || "application/octet-stream" });
    });
  }
  function mensagemDeFalha(falha) {
    if (falha && falha.name === "TypeError") return MSG_REDE;
    return falha && falha.message ? falha.message : "Não foi possível ler o e-mail.";
  }

  function el(tag, classe, texto) {
    var elemento = document.createElement(tag);
    if (classe) elemento.className = classe;
    if (texto !== undefined && texto !== null) elemento.textContent = texto;
    return elemento;
  }

  // Pelo nome; sem extensão (o celular às vezes manda "document"), pelo tipo.
  var TIPOS = { "application/pdf": ".pdf", "message/rfc822": ".eml", "text/plain": ".txt", "application/vnd.ms-outlook": ".msg" };
  function extensaoDe(arquivo) {
    var nome = (arquivo.name || "").toLowerCase();
    var ponto = nome.lastIndexOf(".");
    if (ponto !== -1 && ponto < nome.length - 1) return nome.slice(ponto);
    return TIPOS[(arquivo.type || "").toLowerCase()] || "";
  }

  function iniciar(bloco) {
    var url = bloco.getAttribute("data-url");
    if (!url) return;
    var maximo = parseInt(bloco.getAttribute("data-max-bytes"), 10) || 0;
    var entrada = bloco.querySelector("[data-tri-arquivo]");
    var botaoColar = bloco.querySelector("[data-tri-colar]");
    var painelTexto = bloco.querySelector("[data-tri-texto]");
    var campoTexto = bloco.querySelector("[data-tri-texto-campo]");
    var botaoLerTexto = bloco.querySelector("[data-tri-ler-texto]");
    var estado = bloco.querySelector("[data-tri-estado]");
    var erro = bloco.querySelector("[data-tri-erro]");
    var resultado = bloco.querySelector("[data-tri-resultado]");

    function csrf() {
      var campo = bloco.querySelector('input[name="csrfmiddlewaretoken"]');
      return campo ? campo.value : "";
    }

    function anunciar(texto) { estado.textContent = texto; }

    function mostrarErro(texto) {
      erro.textContent = texto;
      erro.hidden = !texto;
    }

    function ocupado(sim) {
      bloco.classList.toggle("is-lendo", sim);
      bloco.setAttribute("aria-busy", sim ? "true" : "false");
      [entrada, botaoColar, botaoLerTexto].forEach(function (b) { if (b) b.disabled = sim; });
    }

    function descrever(resumo) {
      var partes = [];
      if (resumo.remetente) partes.push("De " + resumo.remetente);
      if (resumo.enviado_em) partes.push("em " + resumo.enviado_em);
      var texto = partes.join(" ");
      if (resumo.assunto) texto += (texto ? ": " : "") + "“" + resumo.assunto + "”";
      if (resumo.quando) texto += " — evento em " + resumo.quando;
      return texto || "E-mail lido.";
    }

    function mostrar(dados) {
      resultado.textContent = "";
      var candidatos = dados.candidatos || [];
      var titulo = el("p", "tri__titulo");
      if (dados.sugerido) {
        var sugerido = candidatos.filter(function (c) { return c.modulo === dados.sugerido; })[0];
        titulo.appendChild(document.createTextNode("Parece um pedido de "));
        titulo.appendChild(el("b", "", sugerido ? sugerido.rotulo : dados.sugerido));
        titulo.appendChild(document.createTextNode(", mas não tenho certeza. Onde registrar?"));
      } else {
        titulo.appendChild(el("b", "", "Não consegui dizer de que módulo é este pedido."));
        titulo.appendChild(document.createTextNode(" Escolha onde registrar:"));
      }
      resultado.appendChild(titulo);
      resultado.appendChild(el("p", "tri__resumo", descrever(dados.resumo || {})));

      var avisos = (dados.resumo && dados.resumo.avisos) || [];
      if (avisos.length) {
        var lista = el("ul", "tri__avisos");
        lista.setAttribute("aria-label", "Avisos da leitura do e-mail");
        avisos.forEach(function (a) { lista.appendChild(el("li", "tri__aviso", a)); });
        resultado.appendChild(lista);
      }

      var opcoes = el("ul", "tri__opcoes");
      candidatos.forEach(function (c) {
        var li = el("li");
        var link = el("a", "tri__opcao" + (c.modulo === dados.sugerido ? " tri__opcao--sugerida" : ""));
        link.href = c.url;
        link.appendChild(el("span", "", c.rotulo));
        var detalhe = c.sinais && c.sinais.length ? c.sinais.slice(0, 3).join(", ") : "sem sinais no texto";
        if (c.confianca) detalhe = Math.round(c.confianca * 100) + "% · " + detalhe;
        link.appendChild(el("small", "", detalhe));
        li.appendChild(link);
        opcoes.appendChild(li);
      });
      resultado.appendChild(opcoes);
      resultado.hidden = false;
      anunciar("");
    }

    function enviar(corpo) {
      mostrarErro("");
      resultado.hidden = true;
      ocupado(true);
      anunciar("Lendo o e-mail…");
      fetch(url, {
        method: "POST",
        body: corpo,
        credentials: "same-origin",
        headers: { "X-CSRFToken": csrf(), "X-Requested-With": "XMLHttpRequest" }
      })
        .then(function (resposta) {
          var tipo = resposta.headers.get("Content-Type") || "";
          if (tipo.indexOf("application/json") === -1) {
            throw new Error(
              resposta.redirected ? "Sua sessão expirou. Entre de novo e solte o e-mail outra vez." :
              resposta.status === 413 ? "O arquivo é grande demais para enviar." :
              "Não foi possível ler o e-mail agora (erro " + resposta.status + ")."
            );
          }
          return resposta.json().then(function (dados) {
            if (!resposta.ok) throw new Error(dados.erro || "Não foi possível ler o e-mail.");
            return dados;
          });
        })
        .then(function (dados) {
          var candidatos = dados.candidatos || [];
          var sugerido = candidatos.filter(function (c) { return c.modulo === dados.sugerido; })[0];
          if (dados.decidir && sugerido && sugerido.url) {
            anunciar("Parece um pedido de " + sugerido.rotulo + ": abrindo a tela já preenchida…");
            window.location.assign(sugerido.url);
            return;
          }
          if (!candidatos.length) {
            throw new Error("Você não tem acesso a nenhum módulo que registre pedidos por e-mail.");
          }
          mostrar(dados);
          if (painelTexto && !painelTexto.hidden) alternarTexto(false);
        })
        .catch(function (falha) {
          anunciar("");
          mostrarErro(mensagemDeFalha(falha));
        })
        .then(function () { ocupado(false); });
    }

    function lerArquivo(arquivo) {
      if (!arquivo) return;
      if (EXTENSOES.indexOf(extensaoDe(arquivo)) === -1) {
        mostrarErro("Envie o e-mail em .eml, .msg, .pdf ou .txt — ou cole o texto.");
        return;
      }
      if (maximo && arquivo.size > maximo) {
        mostrarErro("O arquivo passa do limite de " + Math.round(maximo / 1048576) + " MB.");
        return;
      }
      ocupado(true);
      anunciar("Abrindo o arquivo…");
      copiaDe(arquivo).then(function (copia) {
        var dados = new FormData();
        dados.append("arquivo", copia, copia.name);
        enviar(dados);
      }, function () {
        ocupado(false);
        anunciar("");
        mostrarErro(MSG_LEITURA);
      });
    }

    function lerTexto() {
      var texto = campoTexto ? campoTexto.value.trim() : "";
      if (!texto) {
        mostrarErro("Cole o texto do e-mail antes de ler.");
        if (campoTexto) campoTexto.focus();
        return;
      }
      var dados = new FormData();
      dados.append("texto", texto);
      enviar(dados);
    }

    function alternarTexto(abrir) {
      if (!painelTexto || !botaoColar) return;
      painelTexto.hidden = !abrir;
      botaoColar.setAttribute("aria-expanded", abrir ? "true" : "false");
      if (abrir && campoTexto) campoTexto.focus();
    }

    if (entrada) {
      entrada.addEventListener("change", function () {
        lerArquivo(entrada.files && entrada.files[0]);
        entrada.value = "";
      });
    }
    if (botaoColar) botaoColar.addEventListener("click", function () { alternarTexto(painelTexto.hidden); });
    if (botaoLerTexto) botaoLerTexto.addEventListener("click", lerTexto);
    if (campoTexto) {
      campoTexto.addEventListener("keydown", function (evento) {
        if (evento.key === "Enter" && (evento.ctrlKey || evento.metaKey)) {
          evento.preventDefault();
          lerTexto();
        }
      });
    }

    function temArquivos(evento) {
      var tipos = evento.dataTransfer && evento.dataTransfer.types;
      return Boolean(tipos) && Array.prototype.indexOf.call(tipos, "Files") !== -1;
    }
    var profundidade = 0;
    bloco.addEventListener("dragenter", function (evento) {
      if (!temArquivos(evento)) return;
      evento.preventDefault();
      profundidade += 1;
      bloco.classList.add("is-arrastando");
    });
    bloco.addEventListener("dragover", function (evento) {
      if (!temArquivos(evento)) return;
      evento.preventDefault();
      evento.dataTransfer.dropEffect = "copy";
    });
    bloco.addEventListener("dragleave", function (evento) {
      if (!temArquivos(evento)) return;
      profundidade = Math.max(0, profundidade - 1);
      if (!profundidade) bloco.classList.remove("is-arrastando");
    });
    bloco.addEventListener("drop", function (evento) {
      if (!temArquivos(evento)) return;
      evento.preventDefault();
      profundidade = 0;
      bloco.classList.remove("is-arrastando");
      var arquivos = evento.dataTransfer && evento.dataTransfer.files;
      if (arquivos && arquivos.length > 1) {
        mostrarErro("Solte um e-mail por vez.");
        return;
      }
      lerArquivo(arquivos && arquivos[0]);
    });
  }

  document.querySelectorAll("[data-triagem-email]").forEach(iniciar);
})();
